src/components/panel/panel.js

```diff
@@ -1636,6 +1685,91 @@ MdPanelRef.prototype._animateClose = function() {
   });
 };
 
+/**
+ * Registers a listener with the panel. The callback should return a promise,
+ * which will allow the action to continue when it gets resolved, or will
+ * prevent an action if it is rejected.
+ * @param {string} type Type of listener.
+ * @param {!angular.$q.Promise<!any>} callback Callback to be registered.
+ * @returns {!MdPanelRef}
+ */
+MdPanelRef.prototype.registerListener = function(type, callback) {
+  var error = null;
+
+  if (!angular.isString(type)) {
+    error = 'Listener type must be a string, instead got ' + typeof type;
+  } else if (!angular.isFunction(callback)) {
+    error = 'Listener callback must be a function, instead got ' + typeof callback;
+  }
+
+  if (error) {
+    throw new Error(error);
+  }
+
+  var listeners = this._listeners[type] = this._listeners[type] || [];
+
+  if (listeners.indexOf(callback) === -1) {
+    listeners.push(callback);
+  }
+
+  return this;
+};
+
+/**
+ * Removes a registered listener.
+ * @param {string} type Type of listener to be removed.
+ * @param {Function} callback Listener to be removed.
+ * @returns {!MdPanelRef}
+ */
+MdPanelRef.prototype.removeListener = function(type, callback) {
+  var index = this._listeners[type] ?
+    this._listeners[type].indexOf(callback) : -1;
+
+  if (index > -1) {
+    this._listeners[type].splice(index, 1);
+  }
+
+  return this;
+};
+
+
+/**
+ * Removes all listeners.
+ * @param {string=} type Type of listeners to be removed.
+ * If ommited, all listeners types will be removed.
+ * @returns {!MdPanelRef}
+ */
+MdPanelRef.prototype.removeAllListeners = function(type) {
+  if (type) {
+    this._listeners[type] = [];
+  } else {
+    this._listeners = Object.create(null);
+  }
+
+  return this;
+};
+
+
+/**
+ * Invokes all the listeners from a specific type.
+ * @param {string} type Type of listeners to be invoked.
+ * @returns {!angular.$q.Promise<!MdPanelRef>}
+ * @private
+ */
+MdPanelRef.prototype._callListeners = function(type) {
+  var listeners = this._listeners && this._listeners[type];
+
+  if (listeners && listeners.length) {
+    var invokedListeners = listeners.map(function(currentListener) {
+      return currentListener(this);
+    }, this);
+
+    return this._$q.all(invokedListeners);
```

Comment:
Reversing them before calling shouldn't be a big deal, however I don't know if we should count on the order being correct, because they're using promises and that implies that it's async.